### PR TITLE
Add bridged checker

### PR DESCRIPTION
**Pull request by kiritofeng**
Replaces the more ad-hoc `testlib` checkers with a checker module.

### CODE REVIEW

File: dmoj/checkers/bridged.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+def get_executor(checker_kwargs, problem_id):
+    global executor
+
+    if executor is None:
+        if 'files' not in checker_kwargs:
+            raise InternalError('No checker file[s] specified!')
+        if 'lang' not in checker_kwargs:
+            raise InternalError('Language not specified for checker!')
+
+        filenames = list(map(lambda x: os.path.join(get_problem_root(problem_id), x), checker_kwargs['files']))
```
Comment (kiritofeng) [inline dmoj/checkers/bridged.py]: Given this shares a lot of logic with the generator, should we move this into a `dmoj.utils` file?

File: dmoj/checkers/bridged.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    checker_kwargs.update(checker_defaults)
+    executor = get_executor(checker_kwargs, problem_id)
+
+    with mktemp(judge_input) as arg1, mktemp(process_output) as arg2, mktemp(judge_output) as arg3:
+        process = executor.launch(arg1.name, arg2.name, arg3.name, memory=checker_kwargs['memory_limit'],
+                                  time=checker_kwargs['time_limit'])
+
+        proc_output, error = map(utf8text, process.communicate())
+
+        # We use the testlib.h return codes
```
Comment (Xyene) [inline dmoj/checkers/bridged.py]: Then make these constants

File: dmoj/checkers/bridged.py (hunk 1 above)
Comment (Xyene) [reply dmoj/checkers/bridged.py]: yes

File: dmoj/checkers/bridged.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        clazz = clazz.Executor
+
+        fs = clazz.fs + [tempfile.gettempdir()]
+        clazz = type('Executor', (clazz,), {'fs': fs})
+
+        if issubclass(clazz, CompiledExecutor):
+            compiler_time_limit = checker_kwargs['compiler_time_limit'] or clazz.compiler_time_limit
+            clazz = type('Executor', (clazz,), {'compiler_time_limit': compiler_time_limit})
+
+        if hasattr(clazz, 'flags'):
```
Comment (Xyene) [inline dmoj/checkers/bridged.py]: Note that since we removed Windows support, this hack is no longer necessary.

File: dmoj/checkers/bridged.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            # Optimize the common case.
+            if use_c_or_cpp:
+                # Some checkers (like those using testlib.h) take an extremely long time
+                # to compile, so we cache them.
+                executor = clazz('_checker', None, aux_sources=sources, cached=True)
+            else:
+                if len(sources) > 1:
+                    raise InternalError('non-C/C++ checker cannot be multi-file')
+                executor = clazz('_checker', list(sources.values())[0])
+        except CompileError as err:
```
Comment (Xyene) [inline dmoj/checkers/bridged.py]: This branch is no longer necessary after #489, you can just let the exception bubble up.

File: dmoj/utils/aux_files.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import os
+import tempfile
+import traceback
+
+from dmoj.error import InternalError
+from dmoj.executors import executors
+from dmoj.executors.compiled_executor import CompiledExecutor
+
+
+def _mktemp(data):
```
Comment (quantum5) [inline dmoj/utils/aux_files.py]: If you are going to export this, why does it have an `_` prefix?

File: dmoj/utils/aux_files.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            if grader in executors:
+                return grader
+        return None
+
+    use_cpp = any(map(lambda name: os.path.splitext(name)[1] in ['.cpp', '.cc'], filenames))
+    if lang is None:
+        best_choices = ('CPP17', 'CPP14', 'CPP11', 'CPP0X', 'CPP03') if use_cpp else ('C11', 'C')
+        lang = find_runtime(best_choices)
+
+    clazz = executors.get(lang)
```
Comment (quantum5) [inline dmoj/utils/aux_files.py]: It is obvious that this is a module, so why `clazz`?

File: dmoj/utils/aux_files.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    if not clazz:
+        raise IOError('could not find an appropriate C++ executor')
+
+    clazz = clazz.Executor
```
Comment (quantum5) [inline dmoj/utils/aux_files.py]: I think this is better called `Executor`. If you really like to call it a class, use `cls`.

File: dmoj/utils/aux_files.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        compiler_time_limit = compile_time_limit or clazz.compiler_time_limit
+        clazz = type('Executor', (clazz,), {'compiler_time_limit': compiler_time_limit})
+
+    # Optimize the common case.
+    if use_cpp:
+        # Some auxilary files (like those using testlib.h) take an extremely long time to compile, so we cache them.
+        executor = clazz('_aux_file', None, aux_sources=sources, cached=True)
+    else:
+        if len(sources) > 1:
+            raise InternalError('non-C/C++ auxilary programs cannot be multi-file')
```
Comment (quantum5) [inline dmoj/utils/aux_files.py]: So you can't use `C` with multi file here?

File: dmoj/checkers/bridged.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        filenames = list(map(lambda x: os.path.join(get_problem_root(problem_id), x), checker_kwargs['files']))
+        lang = checker_kwargs['lang']
+        executor = compile_aux_files(filenames, lang, checker_kwargs['compiler_time_limit'])
+
+    return executor
+
+
+def check(process_output, judge_output, judge_input, checker_kwargs,
+          problem_id, point_value=None, **kwargs) -> Optional[CheckerResult]:
+    checker_kwargs.update(checker_defaults)
```
Comment (quantum5) [inline dmoj/checkers/bridged.py]: What? Are they defaults or overrides?

File: dmoj/checkers/bridged.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        WA = 1
+        PE = 2
+        IE = 3
+
+        if process.returncode == AC:
+            if checker_kwargs['feedback']:
+                return CheckerResult(True, point_value, feedback=proc_output)
+            else:
+                return CheckerResult(True, point_value)
+        elif process.returncode in [WA, PE]:
```
Comment (quantum5) [inline dmoj/checkers/bridged.py]: Standard way is with a tuple.

File: dmoj/checkers/bridged.py (hunk 2 above)
Comment (Xyene) [reply dmoj/checkers/bridged.py]: I meant somewhere in global scope

File: dmoj/checkers/bridged.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+        filenames = list(map(lambda x: os.path.join(get_problem_root(problem_id), x), checker_kwargs['files']))
+        lang = checker_kwargs['lang']
+        executor = compile_aux_files(filenames, lang, checker_kwargs['compiler_time_limit'])
+
+    return executor
+
+
+def check(process_output, judge_output, judge_input, checker_kwargs,
+          problem_id, point_value=None, **kwargs) -> Optional[CheckerResult]:
```
Comment (Xyene) [inline dmoj/checkers/bridged.py]: wait why is the `CheckerResult` optional?

File: dmoj/checkers/bridged.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+    # Update checker_kwargs with defaults
+    for key, value in checker_defaults.items():
+        checker_kwargs.setdefault(key, value)
+
+    executor = get_executor(checker_kwargs, problem_id)
+
+    with mktemp(judge_input) as arg1, mktemp(process_output) as arg2, mktemp(judge_output) as arg3:
+        process = executor.launch(arg1.name, arg2.name, arg3.name, memory=checker_kwargs['memory_limit'],
```
Comment (Xyene) [inline dmoj/checkers/bridged.py]: don't you need to pipe output and error?

File: dmoj/utils/aux_files.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
+import os
+import tempfile
+import traceback
+
+from dmoj.error import InternalError
+from dmoj.executors import executors
+from dmoj.executors.compiled_executor import CompiledExecutor
+
+
+def mktemp(data):
```
Comment (Xyene) [inline dmoj/utils/aux_files.py]: _mktemp

Comment (quantum5) [reply dmoj/utils/aux_files.py]: This is exported, so no _.

File: dmoj/checkers/bridged.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+executor = None
+
+
+def get_executor(checker_kwargs, problem_id):
+    global executor
+
+    if executor is None:
+        if 'files' not in checker_kwargs:
+            raise InternalError('No checker file[s] specified!')
```
Comment (Xyene) [inline dmoj/checkers/bridged.py]: our error messages are usually lowercase, also (2)

File: dmoj/checkers/bridged.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+def check(process_output, judge_output, judge_input, checker_kwargs,
+          problem_id, point_value=None, **kwargs) -> CheckerResult:
+
+    # Update checker_kwargs with defaults
+    for key, value in checker_defaults.items():
+        checker_kwargs.setdefault(key, value)
+
+    executor = get_executor(checker_kwargs, problem_id)
+
+    with mktemp(judge_input) as arg1, mktemp(process_output) as arg2, mktemp(judge_output) as arg3:
```
Comment (Xyene) [inline dmoj/checkers/bridged.py]: give these better names

File: dmoj/utils/aux_files.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+def compile_aux_files(filenames, lang=None, compiler_time_limit=None):
+    sources = {}
+
+    try:
+        for filename in filenames:
+            with open(filename, 'rb') as f:
+                sources[os.path.basename(filename)] = f.read()
+    except Exception:
+        traceback.print_exc()
+        raise IOError('Could not read source!')
```
Comment (Xyene) [inline dmoj/utils/aux_files.py]: you probably want to just... let this exception happen so it gets collected by the IE handler

**Pull request merged**